File: chrome/browser/commerce/shopping_list/android/shopping_data_provider_bridge.cc
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <vector>

#include "base/android/jni_android.h"
#include "base/android/jni_array.h"
#include "chrome/browser/commerce/shopping_list/jni_headers/ShoppingDataProviderBridge_jni.h"
#include "chrome/browser/commerce/shopping_list/shopping_data_provider.h"
#include "components/power_bookmarks/core/proto/power_bookmark_meta.pb.h"
#include "content/public/browser/web_contents.h"

using base::android::JavaParamRef;
using base::android::ScopedJavaLocalRef;

namespace shopping_list {

namespace android {

ScopedJavaLocalRef<jbyteArray> JNI_ShoppingDataProviderBridge_GetForWebContents(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_web_contents) {
  content::WebContents* web_contents =
      content::WebContents::FromJavaWebContents(j_web_contents);
  DCHECK(web_contents);

  shopping_list::ShoppingDataProvider* data_provider =
      shopping_list::ShoppingDataProvider::FromWebContents(web_contents);

  if (!data_provider)
    return nullptr;

  std::unique_ptr<power_bookmarks::PowerBookmarkMeta> meta =
      data_provider->GetCurrentMetadata();

  if (!meta)
    return nullptr;

  int size = meta->ByteSize();
  std::vector<uint8_t> data(size);
  meta->SerializeToArray(data.data(), size);
  return base::android::ToJavaByteArray(env, data.data(), size);
}

}  // namespace android

}  // namespace shopping_list

File: chrome/browser/commerce/shopping_list/android/java/src/org/chromium/chrome/browser/commerce/shopping_list/ShoppingDataProviderBridge.java
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.commerce.shopping_list;

import androidx.annotation.Nullable;

import com.google.protobuf.InvalidProtocolBufferException;

import org.chromium.base.Log;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.chrome.browser.power_bookmarks.PowerBookmarkMeta;
import org.chromium.content_public.browser.WebContents;

/** A bridge for accessing the native ShoppingDataProvider from java. */
@JNINamespace("shopping_list::android")
public class ShoppingDataProviderBridge {
    private static final String TAG = "ShoppingDataProvider";

    /**
     * Gets the shopping data for the last navigation of the provided web contents if it exists.
     *
     * @param webContents The web contents to get the shopping data for.
     * @return An object containing the shopping data or null.
     */
    public static @Nullable PowerBookmarkMeta getForWebContents(WebContents webContents) {
        if (webContents == null) return null;

        byte[] protoBytes = ShoppingDataProviderBridgeJni.get().getForWebContents(webContents);
        if (protoBytes == null) return null;

        try {
            return PowerBookmarkMeta.parseFrom(protoBytes);
        } catch (InvalidProtocolBufferException ex) {
            Log.w(TAG, "Failed to parse shopping data: " + ex.getMessage());
            return null;
        }
    }

    @NativeMethods
    interface Natives {
        byte[] getForWebContents(WebContents webContents);
    }
}
